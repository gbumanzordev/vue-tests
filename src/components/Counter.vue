<template>
  <h2>{{ customTitle }}</h2>
  <p>{{ count }}<sup>2</sup> = {{ squareCounter }}</p>
  <p data-test-id="counter">{{ count }}</p>
  <div class="buttons">
    <button @click="increaseCounter()">+ 1</button>
    <button @click="decreaseCounter()">- 1</button>
  </div>
</template>

<script>
export default {
  name: 'Counter',
  props: {
    title: String,
    start: {
      type: Number,
      default: 25,
      validator(start) {
        return start > 0;
      },
    },
  },
  data() {
    return {
      count: this.start ?? 1,
    };
  },
  methods: {
    increaseCounter() {
      this.count++;
    },
    decreaseCounter() {
      this.count--;
    },
  },
  computed: {
    squareCounter() {
      return this.count * this.count;
    },
    customTitle() {
      return this.title || 'Counter';
    },
  },
};
</script>

<style scoped>
button {
  background-color: #64bb87;
  border-radius: 5px;
  border: 1px solid white;
  color: white;
  cursor: pointer;
  margin: 0 5px;
  padding: 15px;
  transition: 0.3s ease-in-out;
}

button:hover {
  background-color: #5aa67b;
  transition: 0.3s ease-in-out;
}

button + button {
  margin-left: 8px;
}
</style>
