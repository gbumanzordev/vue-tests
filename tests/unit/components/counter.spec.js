import { shallowMount } from '@vue/test-utils';
import Counter from '@/components/Counter';
import { findByDataTestId } from '../../utils/test-utils';

describe('Counter Component', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(Counter);
  });

  test('should match snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('should render h2 with default value', () => {
    expect(wrapper.find('h2').exists()).toBeTruthy();
    const headerValue = wrapper.find('h2').text();

    expect(headerValue).toBe('Counter');
  });

  test('should render default value (25) in paragraph', () => {
    const counterParagraph = wrapper.find(findByDataTestId('counter'));
    expect(counterParagraph.text()).toBe('25');
  });

  test('should increase and decrease with one the counter value', async () => {
    const [increaseBtn, decreaseBtn] = wrapper.findAll('button');

    await increaseBtn.trigger('click');
    await increaseBtn.trigger('click');
    await increaseBtn.trigger('click');

    await decreaseBtn.trigger('click');
    await decreaseBtn.trigger('click');

    const counterParagraph = wrapper.find(findByDataTestId('counter'));
    expect(counterParagraph.text()).toEqual('26');
  });

  test('should set default value', () => {
    const { start } = wrapper.props();
    const printedValue = wrapper.find(findByDataTestId('counter'));

    expect(start).toEqual(25);
    expect(printedValue.text()).toEqual(start.toString());
  });

  test('should show the prop title', () => {
    const title = 'Hello World';
    const customWrapper = shallowMount(Counter, {
      props: {
        title,
      },
    });

    expect(customWrapper.find('h2').text()).toBe(title);
  });
});
