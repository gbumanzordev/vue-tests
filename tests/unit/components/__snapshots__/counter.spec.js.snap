// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Component should match snapshot 1`] = `
<h2>Counter</h2>
<p>25<sup>2</sup> = 625</p>
<p data-test-id="counter">25</p>
<div class="buttons"><button>+ 1</button><button>- 1</button></div>
`;
